orangecontrib/bioinformatics/annotation/annotate_samples.py

```diff
@@ -0,0 +1,174 @@
+
+        # compute U, mu, sigma
+        n = data_ge_ranked.shape[0]
+        n2 = n - 1
+        u = data_ge_ranked - 1
+        mu = n2 / 2
+        sigma = np.zeros(data_ge_ranked.shape[1])
+        for i in range(data_ge_ranked.shape[1]):
+            _, counts = np.unique(data_ge_ranked[:, i], return_counts=True)
+            sigma[i] = np.sqrt(
+                1 * n2 / 12 * ((n + 1) - np.sum((counts ** 3 - counts)) /
+                               (n * (n - 1))))
+
+        # compute z
+        z = (u - mu) / (sigma + 1e-16)
+
+        # gene selection
+        attributes_np = np.array([
+            a.attributes.get("Entrez ID") for a in data.domain.attributes])
+        attributes_sets = [set(attributes_np[row > 1]) - {None} for row in z]
+        return attributes_sets, z
+
+    @staticmethod
+    def _group_marker_attributes(markers):
+        """
+        Function transforms annotations table to dictionary with format
+        {annotation1: [attributes], annotation2: [attributes], ...}
+        """
+        types_dict = defaultdict(set)
+        for m in markers:
+            types_dict[str(m["Cell Type"])].add(int(m["Entrez ID"].value))
+        return types_dict
+
+    def _assign_annotations(self, items_sets, available_annotations):
+        """
+        Function get set of attributes (e.g. genes) that represents each item
+        and attributes for each annotation. It returns the annotations most
+        significant for each cell.
+
+        Parameters
+        ----------
+        items_sets : list of sets
+            Set of most important attributes for each item.
+        available_annotations : Orange.data.Table
+            Available annotations (e.g. cell types)
+
+        Returns
+        -------
+        ndarray
+            Annotation probabilities
+        list
+            Annotation list
+        """
+        N = 20365  # number of all genes for human
+
+        grouped_annotations = self._group_marker_attributes(
+            available_annotations)
+        grouped_annotations_items = list(grouped_annotations.items())
+
+        def hg_cell(item_attributes):
+            p_values = []
+            prob = np.zeros(len(grouped_annotations_items))
+            for i, (ct, attributes) in enumerate(grouped_annotations_items):
+                x = len(item_attributes & attributes)
+                k = len(item_attributes)  # drawn balls - expressed for item
+                m = len(attributes)  # marked balls - items for a process
+
+                p_value = self.p_value_fun(x, N, m, k)
+
+                p_values.append(p_value)
+                prob[i] = x / (m + 1e-16)
+
+            fdrs = statistics.FDR(p_values)
+            prob[np.array(fdrs) > self.p_threshold] = 0
+            return prob
+
+        return np.array([hg_cell(cg) for cg in items_sets]), \
+            [x[0] for x in grouped_annotations_items]
+
+    def annotate_samples(self, data, available_annotations):
```

Comment:
Add an option into method header to return only cell types present in at least one sample (columns in `scores_table`). Make it default.